Paula comments

javierorman opened the pull request.

javierorman committed:
Updated .gitignore

javierorman committed:
Removed .DS_Store

javierorman committed:
tab on app.py, client.py, script.py, spotify_auth.py

FiddlersCode commented on app.py:
```diff
@@ -1,54 +1,55 @@
-
-if __name__ == '__main__':
-    app.run(debug=True)
+    import os 
```
Test to confirm I have permissions to leave comments.

FiddlersCode commented on script.py:
```diff
@@ -1,105 +1,105 @@
-import os
-import psycopg2
+    import os
```
File needs a descriptive name. Also it's doing too many things. I'd suggest splitting it, perhaps into:
`DBClient` which ONLY handles connections and CRUD (create, read, update, delete) operations with the db
`Another file I haven't worked out yet`

And regarding testing: you could look at an in-memory database to use. I found this https://pydblite.readthedocs.io/en/latest/ which on a brief read looks promising. In-memory DBs are great because you connect to them instead of the real DB and you can make calls to that (and also seed it so e.g. if you want to test an update method, you can start off by inserting something). As it's in-memory, once the process calling it (in your case, the test) finishes, the memory will be released back to the OS and the data will vanish.

javierorman replied:
Looking into in-memory databases tomorrow!

FiddlersCode commented on script.py:
```diff
@@ -1,105 +1,105 @@
-def get_playlist_id(playlist_uri):
-    playlist_id = playlist_uri.split(':')[2]
-    return playlist_id
+    def get_playlist_id(playlist_uri):
```
This doesn't make sense in a DB class/file. It belongs somewhere that's looking at playlist uris and parsing them.

javierorman replied:
I put all API-related methods in the SpotifyAPI class, so it's more organized. Thank you for that suggestion. Now I have to streamline all API methods, which I think is going to clean up the code quite a bit.

FiddlersCode commented on script.py:
```diff
@@ -1,105 +1,105 @@
-def get_tracks_from_playlist(client, playlist_id):
-    pl_tracks = client.get_playlist_tracks(playlist_id)
-    return pl_tracks
+    def get_tracks_from_playlist(client, playlist_id):
```
Same with this - unrelated to DB stuff.

javierorman replied:
I decided to get rid of these one-line functions that all they do is use a method that already exists.

FiddlersCode commented on script.py:
```diff
@@ -1,105 +1,105 @@
-        track_dict['uri'] = item['track']['uri']
-        list_of_tracks.append(track_dict)
-    return list_of_tracks
+    def make_list_of_tracks(playlist_tracks):
```
Same with this - unrelated to DB stuff.

javierorman replied:
Yep.

FiddlersCode commented on script.py:
```diff
@@ -1,105 +1,105 @@
-        mode = track_features['mode']
-        connect_db.cur.execute(f"""INSERT INTO orig_playlist
-                VALUES ({track_uri!r}, {key}, {mode});""")
+    def get_track_info_insert_into_table(list_of_tracks, client):
```
This is doing 2 things - 
1. Parsing some info to get track stuff
2. Inserting into db

Need to separate the two. DB connections (especially when dealing with production databases) should be isolated AF to avoid the risk of mucking it up. 

You need:
`def get_track_info(track_list, audio_features` - don't call `get_audio_features` in this method, pass it in as a param instead (makes it much easier to test / debug / change).

`def insert_into_table(track_info)`
And all this does it insert into dB.

Also, I'm not sure how Python handles db stuff - no need to worry too much at the moment if it's working but at some point in other projects you'll need to think about how many database connections you can have at once, whether they time out, whether you create them or simply grab them from a pool.

FiddlersCode commented on script.py:
```diff
@@ -1,105 +1,105 @@
-    modes = ['Minor', 'Major']
-    return f"{keys[key]} {modes[mode]}"
+    """
+    Spotify represents keys (C, C#/Db, D, etc.) and modes (Minor, Major) with numbers:
```
You could codify this behaviour with a class or enum or something called `SpotifyKeyMap` and have a dictionary / hash / key-value pair that maps the two.

javierorman replied:
I looked into it and I like those ideas, but they always end up taking more code.

FiddlersCode commented on script.py:
```diff
@@ -1,105 +1,105 @@
-    return track_uris
+    def list_track_uris_by_key_and_mode(key, mode):
+        track_uris = []
+        connect_db.cur.execute(f"""SELECT track_uri
```
Again, isolate db methods. Basically, grab the information you need from the db and then use a different method to parse it.

Also remember that DB queries can be potentially expensive in terms of resources so if you write production code that calls a DB too often, it could have a huge performance impact. Basically, get in the habit of realllly isolating DB stuff.

javierorman replied:
Got it. I'll try to see if I can make less db calls. Right now there are calls happening in every iteration of for loops, which is probably a time-consuming way of doing things.

FiddlersCode commented on script.py:
```diff
@@ -1,105 +1,105 @@
+    def create_new_playlists_add_tracks(client, playlist_name):
+        num_keys = 12
+        num_modes = 2
+        for key in range(num_keys):
```
Nested loops are usually a bad idea for a) performance reasons (not necessarily an issue here but can be elsewhere) and b) understanding what's going on. At the least, I'd extract lines 86-89 to a separate private method.

javierorman replied:
Done!

javierorman replied:
This is now happening in app.py by the way. I used a module called 'product' to create a list of tuples out of the two variables.

FiddlersCode commented on script.py:
```diff
@@ -1,105 +1,105 @@
+                new_playlist_id = client.create_playlist(
+                    pl_name=f"{playlist_name}: {real_key_and_mode(key=key, mode=mode)}")                            
+                response = client.add_songs_to_playlist(new_playlist_id, track_uris)
+                # print(response)    
```
Remove commented code.

javierorman replied:
Will do!

FiddlersCode commented on script.py:
```diff
@@ -1,105 +1,105 @@
-    conn = connect_db.conn
-    cur = connect_db.cur
-    create_table()
+    def script(client, playlist_uri):
```
Rename this method :)

javierorman replied:
I was able to get rid of that function altogether. In general, what would you suggest naming a function (or a file) that hosts most of the functionality and calls that make up the program? Sometimes I have one piece of code that is the 'heart' and I struggle with what to call it because it's in charge of calling so many processes.

javierorman committed:
Created db_client.py with class DBClient to host all database logic. Deleted script.py. Other functionality in script.py was moved to api_client.py and app.py.

javierorman committed:
added __pycache__/ to .gitignore

javierorman committed:
deleted __pycache__ directory

javierorman committed:
1. Change name of api_client.py to spotify_api_client.py
2. Cleaned spotify_api_client: took out comments, created build_get_requests_headers and build_post_requests_headers functions

javierorman committed:
Created main.py to host the central processes of the application. Now, main.py is run as a subprocess from app.py. [Not working]

javierorman committed:
Fixed the subprocess call in app.py and the playlist_uri & access_token retrieval in main.py using sys.argv. [Working]

javierorman committed:
Created README.md (file already existed but was empty), moved images to 'images' folder within 'static'.
